### ISSUE TITLE
Understanding ROIPool

### ISSUE THREAD

**Issue by IssamLaradji** [facebookresearch/maskrcnn-benchmark#328]
## ❓ Questions and Help
Shouldn't `feature_map[:,:,:7,:7].max(1)[0]` be equal to `roi_pool(feature_map[None], bbox_tmp)` in the following case? thanks in advance!

```
import torch
from maskrcnn_benchmark.layers import ROIPool
feature_map = torch.rand(1,256, 80, 160).cuda()
bbox_tmp = torch.FloatTensor([[0, 0, 0, 7, 7]]).cuda()
roi_pool = ROIPool((7,7), spatial_scale=1.0)
roi_pool(feature_map[None], bbox_tmp)                                                                                                                                                                                                               
tensor([[[[0.8403, 0.8633, 0.8704, 0.8704, 0.6135, 0.8561, 0.8561],
          [0.8403, 0.8633, 0.8633, 0.8820, 0.8820, 0.8215, 0.3497],
          [0.8174, 0.7352, 0.8100, 0.8820, 0.8820, 0.8215, 0.9108],
          [0.8174, 0.6335, 0.8100, 0.8100, 0.8296, 0.8296, 0.9108],
          [0.9459, 0.9459, 0.6566, 0.6566, 0.8296, 0.8296, 0.3599],
          [0.9459, 0.9459, 0.8508, 0.6646, 0.7663, 0.7663, 0.6380],
          [0.8416, 0.8508, 0.9013, 0.9013, 0.7663, 0.7663, 0.6380]]]],
       device='cuda:0')

feature_map[:,:,:7,:7].max(1)[0]
tensor([[[0.9961, 0.9978, 0.9998, 0.9906, 0.9947, 0.9965, 0.9963],
         [0.9989, 0.9929, 0.9914, 0.9813, 0.9977, 0.9999, 0.9981],
         [0.9976, 0.9995, 0.9979, 0.9955, 0.9948, 0.9991, 0.9985],
         [0.9992, 0.9991, 0.9988, 0.9970, 0.9972, 0.9950, 0.9986],
         [0.9982, 0.9966, 0.9968, 0.9852, 0.9983, 0.9897, 0.9916],
         [0.9890, 0.9920, 0.9989, 0.9997, 0.9959, 0.9977, 0.9932],
         [0.9985, 0.9861, 0.9852, 0.9946, 0.9983, 0.9993, 0.9971]]],
       device='cuda:0')

```

**Comment by fmassa** [comment]
Question: why did you add an extra dimension to `feature_map` via `feature_map[None]`?

**Comment by IssamLaradji** [comment]
Thanks for the reply. It is because `roi_pool(feature_map[None], bbox_tmp)` returned a shape of `(1,256,7,7)` instead of `(1,7,7)`, but I realized that it is because the pooling is done at the tile level not across the 256 channel dimension. However, I still get different results when I do this,
```
>>> roi_pool(feature_map, bbox_tmp).max(1)[0]
tensor([[[0.9989, 0.9998, 0.9998, 0.9977, 0.9999, 0.9999, 0.9982],
         [0.9995, 0.9995, 0.9979, 0.9977, 0.9999, 0.9999, 0.9992],
         [0.9995, 0.9995, 0.9988, 0.9972, 0.9991, 0.9991, 0.9992],
         [0.9992, 0.9991, 0.9988, 0.9983, 0.9983, 0.9986, 0.9997],
         [0.9982, 0.9989, 0.9997, 0.9997, 0.9983, 0.9977, 0.9997],
         [0.9985, 0.9989, 0.9997, 0.9997, 0.9993, 0.9993, 0.9971],
         [0.9985, 0.9984, 0.9974, 0.9983, 0.9993, 0.9993, 0.9993]]],
       device='cuda:0')
```

**Comment by fmassa** [comment]
So, the roi_pool will take an arbitrary region from the feature map (in this case a 7x7 box starting at 0,0) and pool it so that it has 7x7 features.
Here is what I propose: reduce the number of channels of your input to be only 1, and the dimensions to be something like 10x10, and the size of the pooled region to be 3x3. This will be easier to understand what's going on. Something like
```python
feature_map = torch.rand(1, 1, 10, 10).cuda()
bbox = torch.tensor([0, 0, 0, 5, 5], dtype=torch.float32, device='cuda')
roi_pool = ROIPool((3, 3), spatial_scale=1.0)
print(roi_pool(feature_map, bbox))
```

**Comment by IssamLaradji** [comment]
Thanks for the example, so I tried this `feature_map` that has one channel and has the same size as the output size,
```
feature_map = torch.rand(1, 1, 5, 5).cuda()
bbox = torch.tensor([[0, 0, 0, 5, 5]], dtype=torch.float32, device='cuda')
roi_pool = ROIPool((5, 5), spatial_scale=1.0)
print(roi_pool(feature_map, bbox))
tensor([[[[0.9251, 0.7199, 0.7662, 0.7662, 0.6216],
          [0.9251, 0.7199, 0.7199, 0.9361, 0.9361],
          [0.6901, 0.8800, 0.8800, 0.9361, 0.9361],
          [0.9070, 0.9070, 0.8800, 0.7602, 0.5447],
          [0.9070, 0.9070, 0.7602, 0.7602, 0.3550]]]], device='cuda:0')

print(feature_map)
tensor([[[[0.6853, 0.1465, 0.1397, 0.7662, 0.6216],
          [0.9251, 0.0214, 0.7199, 0.4488, 0.0325],
          [0.6901, 0.2399, 0.0745, 0.2070, 0.9361],
          [0.4316, 0.0058, 0.8800, 0.5155, 0.5447],
          [0.3392, 0.9070, 0.3816, 0.7602, 0.3550]]]], device='cuda:0')

```
I am probably missing something, but I don't see why `roi_pool(feature_map, bbox)` has different values from `feature_map`. I see duplicated values in the `roi_pool` output which means the tile cells are bigger than one element and the max value was taken, which is strange to me since the output size is (5,5) and therefore each cell in the grid has to have one element.

I am assuming it tiles the feature map as follows but with a 5x5 grid instead,

![image](https://user-images.githubusercontent.com/3382128/50975357-5d674680-14bb-11e9-90c9-bc88f819a3d1.png)

Thanks.

**Comment by fmassa** [comment]
Your bounding box is going outside of the image, which is causing it to add some padding while performing the crop.

This gives a zero tensor:
```python
feature_map = torch.rand(1, 1, 5, 5).cuda()
bbox = torch.tensor([[0, 0, 0, 4, 4]], dtype=torch.float32, device='cuda')
roi_pool = ROIPool((5, 5), spatial_scale=1.0)
print(roi_pool(feature_map, bbox) - feature_map)
tensor([[[[0., 0., 0., 0., 0.],
          [0., 0., 0., 0., 0.],
          [0., 0., 0., 0., 0.],
          [0., 0., 0., 0., 0.],
          [0., 0., 0., 0., 0.]]]], device='cuda:0')
```

I'm closing the issue, but let me know if you still have questions

**Comment by IssamLaradji** [comment]
Ah that makes sense! thanks a lot :)
